Only allow users with the CAS3_REFERRER role to interact with Apply / Only allow users with the CAS3_ASSESSOR role to interact with Manage

bryangaledxw opened the pull request:
# Changes in this PR

* This PR restricts manager routes to user with the assessor role, and apply routes to users with the referrer role
* When the user first reaches the service, they will be redirected to either the manage dashboard, or the apply start page

# Release checklist

[Release process documentation](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/edit-v2/4247847062?draftShareId=a1c360ab-bd31-4db1-aae3-7cc002761de9).

## Pre-merge checklist

- [ ] Are any changes required to dependent services for this change to work?
  (eg. CAS API)
- [ ] Have they been released to production already?

## Post-merge checklist

- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to test
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to preprod
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to prod

<!-- Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible. You can monitor deployment failures in CircleCI
itself and application errors are found in
[Sentry](https://ministryofjustice.sentry.io/issues/?project=4504129156218880&referrer=sidebar&statsPeriod=24h).
Both events should be automatically sent to our [Slack
channel](https://mojdt.slack.com/archives/C048BJS7S2F). -->

bryangaledxw committed:
Use service-specific type in UserClient and UserService

bryangaledxw committed:
Update UserServer.getActingUser()

Following CAS1's lead, we update UserService.getActingUser() to return the combined results of fetching user data from the Community Accommodation API and the HMPPS Auth API. Significantly, the object returned by getActingUser() will now contain the user roles we are interested in. We remove getUser() as it is now redundant

bryangaledxw committed:
Remove populateUserRegion()

Enabled by our changes to UserService.getActingUser(), we remove populateUserRegion(), and move its responsibilities into populateCurrentUser(). This saves two largely identical tasks being spread across different pieces of middleware. We take this as an opportunity to add missing testing to populateCurrentUser().

bryangaledxw committed:
Add LandingController

We add LandingController, which will be responsible for redirecting the user to their correct starting page, depending on their role

bryangaledxw committed:
Sort imports and exports

bryangaledxw committed:
Move DashboardController and use LandingController

We reorganise DashboardController to live with the manager controllers, and change its path to /dashboard. We then assign LandingController to the root path, so a user reaching the service domain will be redirected to either the dashboard or the applications start page, depending on their role.

We can now integration test the landing controller. We also update some of our existing integration tests to handle the user reaching the application start page directly after login.

bryangaledxw committed:
Add compose() routing helper

Ahead of adding middleware for restricting routes by user role, we add compose() as a companion to actions(), for when we want to apply particular middleware to a group of routes

bryangaledxw committed:
Add createRoleMiddleware()

We add createRoleMiddleware(), which returns middleware that either allows a user to progress to a given handler, or throws an error if they do not have the required role. We will use this with compose() to restrict our routes by user role

bryangaledxw committed:
Add UnauthorizedError

Ahead of restricting certain endpoints to certain user roles, we add this error class for use when a user attempts to reach an endpoint they do not have permission to access. When this is thrown, the status code means it will recognised by the error handler middleware and the user will be logged out

bryangaledxw committed:
Update integration tests to use user roles

Ahead of restricting certain endpoints to certain user roles, we update our integration tests to be run with a mock user of the appropriate role for that test. Though not yet used, we preemptively add setupTestUserWithoutRole() so we can test behaviour when the application is used by someone who is neither an assessor nor a referrer.

We also set our wiremock test user to have both roles, for ease of manual testing.

bryangaledxw committed:
Apply role middleware

We use our role middleware to restrict our routes by user role

bryangaledxw commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Only allow users with the CAS3_ASSESSOR role to interact with Manage](https://trello.com/c/209Srqfj/1337-only-allow-users-with-the-cas3assessor-role-to-interact-with-manage)

bryangaledxw commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Only allow users with the CAS3_REFERRER role to interact with Apply](https://trello.com/c/jJWy1iBY/1285-only-allow-users-with-the-cas3referrer-role-to-interact-with-apply)

bryangaledxw commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Only allow users with the CAS3_REFERRER role to interact with Apply](https://trello.com/c/jJWy1iBY/1285-only-allow-users-with-the-cas3referrer-role-to-interact-with-apply)

Harriethw commented on server/middleware/roleMiddleware.ts:
```diff
@@ -0,0 +1,16 @@
+const wrapHandler =
+  (handler: RequestHandler, role: UserRole) => async (req: Request, res: Response, next: NextFunction) => {
+    if (res.locals.user.roles.includes(role)) {
+      await handler(req, res, next)
```
minor: the IDE is telling me that `await` has no effect on this function, but haven't got it running to tell locally :)

bryangaledxw replied:
Thanks for the heads up on this - took another look at the code and realised the audit step is happening before the role step, which wasn't what I intended - will need to take another look at things

Harriethw reviewed (approved):
nice one! definitely bookmarking this in case we have to implement something similar :)

bryangaledxw committed:
Update audit middleware

We update our audit middleware to not audit its wrapped handler if that handler returns an error. We want this behaviour as we do not want to record audit events in instances where the user is unable to access an endpoint as they have been logged out

bryangaledxw committed:
Correct test description in audit middleware

tahb reviewed (approved):
LGTM
